src/storage/data_table.cpp

```diff
@@ -658,79 +670,119 @@ void DataTable::VerifyNewConstraint(LocalStorage &local_storage, DataTable &pare
 	local_storage.VerifyNewConstraint(parent, constraint);
 }
 
-bool HasUniqueIndexes(TableIndexList &list) {
-	bool has_unique_index = false;
-	list.Scan([&](Index &index) {
-		if (index.IsUnique()) {
-			has_unique_index = true;
-			return true;
-		}
-		return false;
-	});
-	return has_unique_index;
+static void UpdateDeleteIndex(BoundIndex &delete_index, DataChunk &chunk, optional_ptr<Vector> row_ids,
+                              optional_ptr<DataChunk> delete_chunk) {
+	if (!row_ids) {
+		return;
+	}
+
+	auto &append_chunk = delete_chunk ? *delete_chunk : chunk;
+	if (append_chunk.size() == 0) {
+		return;
+	}
+	auto result = delete_index.Append(append_chunk, *row_ids);
+	if (result.HasError()) {
+		throw InternalException("unexpected constraint violation on delete ART: ", result.Message());
+	}
 }
 
-void DataTable::VerifyUniqueIndexes(TableIndexList &indexes, ClientContext &context, DataChunk &chunk,
-                                    optional_ptr<ConflictManager> conflict_manager) {
-	//! check whether or not the chunk can be inserted into the indexes
-	if (!conflict_manager) {
-		// Only need to verify that no unique constraints are violated
-		indexes.Scan([&](Index &index) {
+void DataTable::VerifyUniqueIndexes(TableIndexList &indexes, optional_ptr<TableIndexList> delete_indexes,
```

Comment:
Maybe we can pass the `LocalStorage` instead of the `optional_ptr<TableIndexList> delete_indexes`, since the delete indexes will always come from the local storage anyway?

Reply:
I can pass the `LocalTableStorage`. This is a `static` function, if I pass the `LocalStorage`, then I also have to pass the respective `DataTable`, or we need to remove the `static`. 🤔